src/verify.rs

```diff
@@ -0,0 +1,615 @@
+        (&G1Affine::generator(), &q_prepared),
+    ]);
+
+    if exp.final_exponentiation() != Gt::identity() {
+        Err(VerificationError::SignatureFailedVerification)
+    } else {
+        Ok(())
+    }
+}
+
+pub fn verify_on_g1(
+    public_key: &[u8],
+    message: &[u8],
+    signature: &[u8],
+    domain_separation_tag: &str,
+) -> Result<(), VerificationError> {
+    let pub_key_bytes: &[u8; 96] = public_key
+        .try_into()
+        .map_err(|_| VerificationError::InvalidPublicKey)?;
+
+    let sig_bytes: &[u8; 48] = signature
+        .try_into()
+        .map_err(|_| VerificationError::InvalidSignatureLength)?;
+
+    let signature_point = G1Affine::from_compressed(sig_bytes).unwrap_or(G1Affine::identity());
+    let pubkey_point = G2Affine::from_compressed(pub_key_bytes).unwrap_or(G2Affine::identity());
+
+    if pubkey_point.is_on_curve().unwrap_u8() != 1 {
+        return Err(VerificationError::InvalidPublicKey);
+    }
+
+    if pubkey_point.is_identity().unwrap_u8() == 1 {
+        return Err(VerificationError::InvalidPublicKey);
+    }
+
+    if message.is_empty() {
+        return Err(VerificationError::EmptyMessage);
+    }
+
+    if signature.is_empty() {
+        return Err(VerificationError::InvalidSignatureLength);
+    }
+
+    let m = <G1Projective as HashToCurve<ExpandMsgXmd<Sha256>>>::hash_to_curve(
+        message,
+        domain_separation_tag.as_bytes(),
+    );
+
+    let pubkey_prepared = G2Prepared::from(pubkey_point.neg());
+    let g2_base = G2Prepared::from(G2Affine::generator());
+    let exp = multi_miller_loop(&[
+        (&G1Affine::from(m), &pubkey_prepared),
+        (&signature_point, &g2_base),
+    ]);
+
+    if exp.final_exponentiation() != Gt::identity() {
+        Err(VerificationError::SignatureFailedVerification)
+    } else {
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::verify::{verify_beacon, Beacon, SchemeID, VerificationError};
+    use bls12_381::{G1Affine, G2Affine};
+
+    #[test]
+    fn default_beacon_verifies() {
+        let public_key = dehexify("88a8227b75dba145599d894d33eebde3b36fef900d456ae2cc4388867adb4769c40359f783750a41b4d17e40f578bfdb");
+        let prev_sig = dehexify("a2237ee39a1a6569cb8e02c6e979c07efe1f30be0ac501436bd325015f1cd6129dc56fd60efcdf9158d74ebfa34bfcbd17803dbca6d2ae8bc3a968e4dc582f8710c69de80b2e649663fef5742d22fff7d1619b75d5f222e8c9b8840bc2044bce");
+
+        let beacon = Beacon {
+            round_number: 397089,
+            randomness: dehexify("cd435675735e459fb4d9c68a9d9f7b719e59e0a9f5f86fe6bd86b730d01fba42"),
+            signature: dehexify("88ccd9a91946bc0bbef2c6c60a09bbf4a247b1d2059522449aa1a35758feddfad85efe818bbde3e1e4ab0c852d96e65f0b1f97f239bf3fc918860ea846cbb500fcf7c9d0dd3d851320374460b5fc596b8cfd629f4c07c7507c259bf9beca850a"),
+            previous_signature: prev_sig,
+        };
+
+        assert!(matches!(
+            verify_beacon(&SchemeID::PedersenBlsChained, &public_key, &beacon),
```

Comment:
I can't understand how this worked given the above check I commented on

Reply:
I think you're mixing up the construction of the message being signed with the randomness derivation from the sig

Reply:
yup